src/array_partition.jl

```diff
@@ -370,3 +370,18 @@ function LinearAlgebra._swap_rows!(B::ArrayPartition, i::Integer, j::Integer)
     B[i], B[j] = B[j], B[i]
     return B
 end
+
+# linalg mul! overloads for ArrayPartition
+function LinearAlgebra.mul!(C::T, A::T, B::AbstractArray) where T<:ArrayPartition
+    @assert length(C.x) == length(A.x)
```

Comment:
According to the docstring, "An assert might be disabled at various optimization levels. Assert should therefore only be used as a debugging tool". Hence maybe change it to a regular check and explicitly throw an error?